### PR TITLE
[Wait for tests to be written] PRO-3267 PRO-3268 Inline array fields, and a bug fix for object field validation

### PR DESCRIPTION
**Pull request by boutell**
## Summary

Support for inline arrays (array fields with `inline: 'true'`, which display inline without the need for a dialog box). The interface is optimized for schemas with a small number of fields per item, frequently just one.

This PR also fixes PRO-3267, so object fields should stop a save operation with a "required" message without first hitting the server and getting a 400 error.

## What are the specific steps to test this change?

You can link it to the testbed's inline-arrays branch. Then go to Topics → Fancy Fields → Ice Cream Flavors

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [X] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [X] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue [hunk 1]
```diff
@@ -27,27 +78,103 @@
...
   name: 'AposInputArray',
+  components: { draggable },
   mixins: [ AposInputMixin ],
+  props: {
+    generation: {
+      type: Number,
+      required: false,
+      default() {
+        return null;
+      }
```
Comment (ETLaurent) [inline modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue:95]:
```suggestion
      default: null
```

I think function is required  for arrays and objects

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue [hunk 2]
```diff
@@ -27,27 +78,103 @@
...
+    const data = {
+      next,
+      items: expandItems(next)
     };
+    return data;
   },
   computed: {
-    editLabel () {
+    listId() {
+      return `sortableList-${(Math.floor(Math.random() * Math.floor(10000)))}`;
```
Comment (ETLaurent) [inline modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue:108]: could we use cuid here?

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue [hunk 3]
```diff
@@ -27,27 +78,103 @@
...
+      if (this.field.schema.length === 1) {
+        return [
+          {
+            ...this.field.schema[0],
+            hideLabel: true
+          }
+        ];
+      } else {
+        return this.field.schema;
+      }
```
Comment (ETLaurent) [inline modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue:133]: no need for `else` thanks to the return above but that's a minor thing

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue [hunk 4]
```diff
@@ -71,7 +195,67 @@ export default {
...
+    getNext() {
+      // Next should consistently be an array.
+      return (this.value && Array.isArray(this.value.data))
+        ? this.value.data : (this.field.def || []);
+    },
+    disableAdd() {
+      return this.field.max && (this.items.length >= this.field.max);
+    },
+    remove(_id) {
+      this.items = this.items.filter(color => color._id !== _id);
```
Comment (ETLaurent) [inline modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue:208]:
```suggestion
      this.items = this.items.filter(item => item._id !== _id);
```

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue [hunk 5]
```diff
@@ -71,7 +195,67 @@ export default {
...
+    // align-self: stretch seems like the answer, but looks bad
+    // once the field adds additional height to itself for an error.
+    // 15px - 7.5px (padding of icons) = 7.5px
+    padding-top: 7.5px;
+  }
+  .apos-input-array-inline-schema-wrapper {
+    flex-grow: 1;
+  }
+  .apos-drag-handle {
+    padding: 7.5px;
```
Comment (ETLaurent) [inline modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue:259]: are we ok with pixels as floats?

**Review by ETLaurent — COMMENTED** [review]
tested on testbed (not on the `inline-arrays` branch because broken), inline arrays seem to  work well, max and def  options work, adding removing and dragging the items work ✅ 

just few comments

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue (hunk 5 above)
Comment (boutell) [reply modules/@apostrophecms/schema/ui/apos/components/AposInputArray.vue:258]: I'm borrowing from existing practice elsewhere, 7.5px appears quite a bit in this situation.

**Comment by boutell** [comment]
I've added @stuartromanek to the reviewers on this ticket. It won't come out in any case until the 9th, so that gives him an opportunity to weigh in on my design choices, as well as any code concerns.

**Review by ETLaurent — DISMISSED** [review]

**Comment by stuartromanek** [comment]
Eyy love this. The first example I threw it on was a little messy in terms of UI
image

Also, w this implementation of our drag-and-drop functionality I'm getting the buggy behavior where the dragged element flys extreme left in a jumpy and confusing way.. we need to fix that.

https://user-images.githubusercontent.com/1889830/199018779-f884d422-e92a-4a81-9f6e-69983cac6c1b.mov

I could see the need for some solution for long schemas, possibly collapsible forms, but i think this feature set should be good for now. 

@boutell do you want to assign to me to work through some UI or design something up for you to implement? Happy to freestyle some adjustments.

**Comment by boutell** [comment]
Freestyle away!

**Comment by boutell** [comment]
We fixed this up nice in subsidiary PRs, but it can't land until tests are written 😢 I'll have to figure out time for that.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,11 +4,13 @@
 
 ### Adds
 
+* You can now set `inline: true` on schema fields of type `array`. This displays a simple editing interface in the context of the main dialog box for the document in question, avoiding the need to open an additional dialog box. Usually best for cases with just one field or just a few. If your array field has a large number of subfields the default behavior (`inline: false`) is more suitable for your needs.
```
Comment (stuartromanek) [inline CHANGELOG.md:7]: note the `alwaysExpand` subprop of `inline` here, or just the real docs?

**Review by stuartromanek — APPROVED** [review]

**Pull request merged**